core/src/main/java/apoc/trigger/TriggerNewProcedures.java

```diff
@@ -0,0 +1,129 @@
+package apoc.trigger;
+
+import apoc.SystemPropertyKeys;
+import apoc.util.Util;
+import org.neo4j.graphdb.Transaction;
+import org.neo4j.kernel.api.procedure.SystemProcedure;
+import org.neo4j.kernel.internal.GraphDatabaseAPI;
+import org.neo4j.logging.Log;
+import org.neo4j.procedure.Admin;
+import org.neo4j.procedure.Context;
+import org.neo4j.procedure.Description;
+import org.neo4j.procedure.Mode;
+import org.neo4j.procedure.Name;
+import org.neo4j.procedure.Procedure;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import static org.neo4j.configuration.GraphDatabaseSettings.SYSTEM_DATABASE_NAME;
+
+
+public class TriggerNewProcedures {
+    // public for testing purpose
+    public static final String TRIGGER_NOT_ROUTED_ERROR = "The procedure should be routed and executed against a writer database";
+    public static final String TRIGGER_BAD_TARGET_ERROR = "Triggers can only be installed on user databases.";
+
+    @Context public GraphDatabaseAPI db;
+    
+    @Context public Log log;
+    
+    @Context public Transaction tx;
+
+    private void checkInSystemWriter() {
+        TriggerHandlerNewProcedures.checkEnabled();
+        // routing check
+        if (!db.databaseName().equals(SYSTEM_DATABASE_NAME) || !Util.isWriteableInstance(db)) {
+            throw new RuntimeException(TRIGGER_NOT_ROUTED_ERROR);
+        }
+    }
+
+    private void checkTargetDatabase(String databaseName) {
+        if (databaseName.equals(SYSTEM_DATABASE_NAME)) {
+            throw new RuntimeException(TRIGGER_BAD_TARGET_ERROR);
+        }
```

Comment:
I think this current check is a good start but I don't think it's sufficient. What we should is that we should also run a query "SHOW DATABASES" and check that the target database actually exists.

The user shouldn't be able to create a trigger for database "foo" if it doesn't exist. This logic was happening in the old procedures although it wasn't explicit. It was happening because in order to use the old triggers you had to connect to the session database and then validate the trigger statement - none of which was possible if the database didn't exist.

PS. You don't need to run the query "SHOW DATABASES" in a new system transaction because you're already in a system transaction when this code is executing.

Reply:
Added